File: Codes/Bingran Li/Tree/107. Binary Tree Level Order Traversal II.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        // solution 1: just like problem 102, plus--- Collections.reverse()
        
        // solution 2: preorder traversal
        
        
        List<List<Integer>> myList = new ArrayList ();
        if(root==null)
            return myList;
            
        traversal(myList,root,0);
        
        Collections.reverse(myList);
        
        return myList;
        
    }
    public void traversal(List<List<Integer>> myList, TreeNode myNode, int dep){
        
       // if(myNode == null) return;
        
        if(dep==myList.size()){
            // first node of each level, create new levList
            List<Integer> levList = new ArrayList<Integer>();
            myList.add(levList);
            
        }
        // add val of myNode to specific levList
        (myList.get(dep)).add(myNode.val);
        
       if(myNode.left!=null)  traversal(myList,myNode.left,dep+1);
       if(myNode.right!=null) traversal(myList,myNode.right,dep+1);
    }
    
    
}

File: Codes/Yangkun Huang/Tree/107. Binary Tree Level Order Traversal II.cpp
vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> v;
        vector<vector<int>> vv;
        queue<TreeNode*> q;
        if(root)
            q.push(root);
        while(!q.empty())
        {
            vector<int> vi;
            for(int i=0;i<q.size();i++)   
            {
                vi.push_back(q.front()->val);
                q.push(q.front());
                q.pop();
            }
            for(int i=0,j=q.size();i<j;i++) //循环过程中q.size()会改变
            {
                if(q.front()->left)
                q.push(q.front()->left);
                if(q.front()->right)
                q.push(q.front()->right);
                q.pop();
            }
            v.push_back(vi);
        }
        for(int i=v.size()-1;i>=0;i--)
        vv.push_back(v[i]);
        return vv;
    }